source/platforms/allwinner-d1/src/main.rs

```diff
@@ -0,0 +1,623 @@
+        }).await;
+
+        // Receiver task
+        // TODO: This Arc shouldn't be necessary
+        // The reception is handled in the interrupt itself.
+        let prod = kernel.heap().allocate_arc(prod).await;
+        UART_RX_PROD.store(prod.leak().as_ptr(), Ordering::Release);
+
+        kernel.with_registry(|reg| {
+            reg.register_konly::<SimpleSerial>(&kprod).unwrap()
+        }).await;
+
+        Ok(())
+    }
+}
+
+// ----
+
+pub struct TimerQueue {
+    _x: (),
+}
+
+pub struct TQPusher {
+    arr: HeapArc<Mutex<HeapArray<Option<(u64, Message<TimerQueue>)>>>>,
+    chan: KConsumer<Message<TimerQueue>>,
+}
+
+pub struct TQClient {
+    hdl: KernelHandle<TimerQueue>,
+    osc: Reusable<Envelope<Result<TQResponse, TQError>>>,
+}
+
+impl TQClient {
+    pub async fn delay_ms(&mut self, ms: u64) {
+        self.hdl.send(
+            TQRequest::DelayMs(ms),
+            ReplyTo::OneShot(self.osc.sender().unwrap()),
+        ).await.unwrap();
+        self.osc.receive().await.unwrap();
+    }
+}
+
+impl TQPusher {
+    pub async fn run(&mut self) {
+        loop {
+            match self.chan.dequeue_async().await {
+                Ok(msg) => {
+                    let now = TICK_MS.load(Ordering::Acquire);
+                    let mut guard = self.arr.lock().await;
+                    let space = guard.iter_mut().find(|i| i.is_none()).unwrap();
+                    let TQRequest::DelayMs(ms) = &msg.msg.body;
+                    let end = ms.wrapping_add(now);
+                    *space = Some((end, msg));
+                },
+                Err(_) => panic!(),
+            }
+        }
+    }
+}
+
+pub struct TQPopper {
+    arr: HeapArc<Mutex<HeapArray<Option<(u64, Message<TimerQueue>)>>>>,
+    wait: HeapArc<WaitCell>,
+}
+
+impl TQPopper {
+    pub async fn run(&mut self) {
+        loop {
+            self.wait.wait().await.unwrap();
+            let mut guard = self.arr.lock().await;
+            let now = TICK_MS.load(Ordering::Acquire);
+
+            // lol. lmao.
+            for item in guard.iter_mut() {
+                match item.take() {
+                    Some((time, msg)) => {
+                        if time <= now {
+                            let resp = msg.msg.reply_with(Ok(TQResponse::Delayed { now }));
+                            msg.reply.reply_konly(resp).await.unwrap();
+                        } else {
+                            *item = Some((time, msg));
+                        }
+                    },
+                    None => {},
+                }
+            }
+        }
+    }
+}
+
+impl TimerQueue {
+    pub async fn register(kernel: &'static Kernel) -> Result<HeapArc<WaitCell>, ()> {
+        let wait = kernel.heap().allocate_arc(WaitCell::new()).await;
+        let (kprod, kcons) = KChannel::new_async(kernel, 32).await.split();
+        let arr = kernel.heap().allocate_array_with(|| None, 128).await;
+        let arr = kernel.heap().allocate_arc(Mutex::new(arr)).await;
```

Comment:
since this is a fixed size array, and `maitake::sync::Mutex::new` is a const fn, couldn't this just be a static rather than an arc?
```rust
static TIMER_QUEUE_ARRAY: Mutex<[Option<(u64, Message<TimerQueue>)>, 128]> = Mutex::new([None; 128]);
```
should work...

Reply:
Also yes!